### PR TITLE
BiCGStab and CG implementations: possible improvements

### PR DESCRIPTION
**Pull request by eebasso**
## Summary
Corrects some possible minor inefficiencies in the implementation of biconjugate gradient stabilized (BiCGStab) and conjugate gradient (CG) methods defined in MLCGSolverT class. The changes have not been tested yet. This should lead to no change in the outcomes, but hopefully will be faster by reducing memory usage and copy operations.

## Additional background
See the attached pdf for a more in-depth explanation and comparison of current vs proposed 
implementations. [BiCGStab and CG algorithm summary.pdf](https://github.com/AMReX-Codes/amrex/files/12835999/BiCGStab.and.CG.algorithm.summary.pdf)

## Checklist
The proposed changes:
- improves computational efficiency

### CODE REVIEW

File: Src/LinearSolvers/MLMG/AMReX_MLCGSolver.H [hunk 1]
```diff
@@ -96,27 +96,22 @@ MLCGSolverT<MF>::solve_bicgstab (MF& sol, const MF& rhs, RT eps_rel, RT eps_abs)
 
     MF ph(ba, dm, ncomp, sol.nGrowVect(), MFInfo(), factory);
     MF sh(ba, dm, ncomp, sol.nGrowVect(), MFInfo(), factory);
-    ph.setVal(RT(0.0));
-    sh.setVal(RT(0.0));
 
     MF sorig(ba, dm, ncomp, nghost, MFInfo(), factory);
     MF p    (ba, dm, ncomp, nghost, MFInfo(), factory);
     MF r    (ba, dm, ncomp, nghost, MFInfo(), factory);
-    MF s    (ba, dm, ncomp, nghost, MFInfo(), factory);
```
Comment (eebasso) [inline Src/LinearSolvers/MLMG/AMReX_MLCGSolver.H]: The variable `s` can be eliminated by using residual `r`

File: Src/LinearSolvers/MLMG/AMReX_MLCGSolver.H [hunk 2]
```diff
@@ -225,7 +215,14 @@ MLCGSolverT<MF>::solve_bicgstab (MF& sol, const MF& rhs, RT eps_rel, RT eps_abs)
         {
             ret = 4; break;
         }
-        rho_1 = rho;
```
Comment (eebasso) [inline Src/LinearSolvers/MLMG/AMReX_MLCGSolver.H:224]: No need to save old rho as we only need rhTv to compute beta

**Comment by WeiqunZhang** [comment]
@eebasso A number of regression tests have failed. Some of them were just roundoff errors, but some produced NaNs.

https://ccse.lbl.gov/pub/GpuRegressionTesting/AMReX/2023-11-01-003/index.html
https://ccse.lbl.gov/pub/RegressionTesting1/AMReX/2023-11-01-001/index.html

If it makes sense, baybe you could split this into a few small PRs.

**Comment by eebasso** [comment]
I think splitting up into multiple PRs is quite reasonable. Shall I update this PR with a smaller change or close it and make a new PR ?

**Comment by WeiqunZhang** [comment]
It might be cleaner to start new PRs.

**Pull request closed without merging**